File: src/audio/titles.ts
export function titleForHum(when: Date, existingTitles: string[]): string {
  const hour = when.getHours()
  const prefixes =
    hour < 5
      ? ['Late Night Jam', 'Night Hum']
      : hour < 12
        ? ['Morning Melody', 'Sunrise Hum']
        : hour < 17
          ? ['Afternoon Sketch', 'Acoustic Idea']
          : hour < 21
            ? ['Evening Hum', 'Golden Hour']
            : ['Late Night Jam', 'Night Hum']

  const prefix = prefixes[0]!
  const used = new Set(existingTitles)
  for (let n = 1; n < 200; n++) {
    const title = `${prefix} #${n}`
    if (!used.has(title)) return title
  }
  return `${prefix} #${Date.now()}`
}

export function formatDuration(ms: number): string {
  const total = Math.max(0, Math.round(ms / 1000))
  const m = Math.floor(total / 60)
  const s = total % 60
  return m > 0 ? `${m}:${s.toString().padStart(2, '0')}` : `0:${s.toString().padStart(2, '0')}`
}

export function formatRecordedAt(iso: string, now = new Date()): string {
  const date = new Date(iso)
  const sameDay =
    date.getFullYear() === now.getFullYear() &&
    date.getMonth() === now.getMonth() &&
    date.getDate() === now.getDate()
  const yesterday = new Date(now)
  yesterday.setDate(now.getDate() - 1)
  const isYesterday =
    date.getFullYear() === yesterday.getFullYear() &&
    date.getMonth() === yesterday.getMonth() &&
    date.getDate() === yesterday.getDate()

  const time = date.toLocaleTimeString(undefined, { hour: 'numeric', minute: '2-digit' })
  if (sameDay) return `Today, ${time}`
  if (isYesterday) return `Yesterday, ${time}`
  return date.toLocaleDateString(undefined, { weekday: 'long', month: 'short', day: 'numeric' })
}

export function formatTimer(ms: number): string {
  const total = Math.floor(ms / 1000)
  const m = Math.floor(total / 60)
  const s = total % 60
  return `${m.toString().padStart(2, '0')}:${s.toString().padStart(2, '0')}`
}

File: Humming/Core/Titles.swift
import Foundation

public enum Titles {
    public static func title(for date: Date, existing: [String]) -> String {
        let hour = Calendar.current.component(.hour, from: date)
        let prefix: String
        switch hour {
        case 0..<5: prefix = "Late Night Jam"
        case 5..<12: prefix = "Morning Melody"
        case 12..<17: prefix = "Afternoon Sketch"
        case 17..<21: prefix = "Evening Hum"
        default: prefix = "Late Night Jam"
        }
        let used = Set(existing)
        for n in 1..<200 {
            let title = "\(prefix) #\(n)"
            if !used.contains(title) { return title }
        }
        return "\(prefix) #\(Int(date.timeIntervalSince1970))"
    }

    public static func duration(_ ms: Double) -> String {
        let total = max(0, Int((ms / 1000).rounded()))
        let m = total / 60
        let s = total % 60
        return m > 0 ? "\(m):\(String(format: "%02d", s))" : "0:\(String(format: "%02d", s))"
    }

    public static func recordedAt(_ date: Date, now: Date = Date()) -> String {
        let calendar = Calendar.current
        let time = date.formatted(date: .omitted, time: .shortened)
        if calendar.isDate(date, inSameDayAs: now) {
            return "Today, \(time)"
        }
        if let yesterday = calendar.date(byAdding: .day, value: -1, to: now),
           calendar.isDate(date, inSameDayAs: yesterday) {
            return "Yesterday, \(time)"
        }
        return date.formatted(.dateTime.weekday(.wide).month(.abbreviated).day())
    }

    public static func timer(_ interval: TimeInterval) -> String {
        let total = Int(interval)
        let m = total / 60
        let s = total % 60
        return String(format: "%02d:%02d", m, s)
    }

    public static func greeting(on date: Date = Date()) -> (kicker: String, line: String) {
        let greetings = [
            ("Yo!", "What's in your head?"),
            ("Haven't seen you in a while", "Feeling creative today?"),
            ("Yo.", "Back for another hit?"),
            ("Welcome back my friend,", "Today is jamming seshhhh"),
        ]
        let day = Int(date.timeIntervalSince1970 / 86_400)
        return greetings[day % greetings.count]
    }
}
